webf/lib/src/html/canvas/canvas_context_2d.dart

```diff
@@ -819,20 +827,90 @@ class CanvasRenderingContext2D extends BindingObject {
     Rect rect = Rect.fromLTWH(x, y, w, h);
     addAction((Canvas canvas, Size size) {
       Paint paint = Paint();
-      if (fillStyle is Color) {
-        paint..color = fillStyle as Color;
-      } else if (fillStyle is CanvasRadialGradient) {
-        paint
-          ..shader =
-              _drawRadialGradient(fillStyle as CanvasRadialGradient, x, y, w, h)
-                  .createShader(rect);
-      } else if (fillStyle is CanvasLinearGradient) {
-        paint
-          ..shader =
-              _drawLinearGradient(fillStyle as CanvasLinearGradient, x, y, w, h)
-                  .createShader(rect);
+      if (fillStyle is Color || fillStyle is CanvasRadialGradient || fillStyle is CanvasLinearGradient) {
+        if (fillStyle is Color) {
+          paint..color = fillStyle as Color;
+        } else if (fillStyle is CanvasRadialGradient) {
+          paint..shader = _drawRadialGradient(fillStyle as CanvasRadialGradient, x, y, w, h).createShader(rect);
+        } else if (fillStyle is CanvasLinearGradient) {
+          paint..shader = _drawLinearGradient(fillStyle as CanvasLinearGradient, x, y, w, h).createShader(rect);
+        }
+        canvas.drawRect(rect, paint);
+      } else if (fillStyle is CanvasPattern) {
+        var canvasPattern = fillStyle as CanvasPattern;
+        if (canvasPattern.image.image_element == null && canvasPattern.image.canvas_element == null) {
+          throw AssertionError('CanvasPattern must be created from a canvas or image');
+        }
+
+        String repetition = canvasPattern.repetition;
+        int patternHeight = 0;
+        int patternWidth = 0;
+        //CanvasPattern created from an image
+        if (canvasPattern.image.image_element != null) {
+          patternHeight = canvasPattern.image.image_element!.height;
+          patternWidth = canvasPattern.image.image_element!.width;
+          switch (repetition) {
```

Comment:
Thanks for your advice.I will optimize the code this weekend. 
And about the cache image with replayAction to draw,if the canvas which i create pattern from is not drawn,
it seems that i can't get the image int the fillRect context, because when the addAction function was called,we just put some  CanvasAction to the _actions,and

Reply:
Thanks for your advice.I will optimize the code this weekend. 
And about the cache image with replayAction to draw,if the CanvasElement which i create pattern from is not drawn, i can't get the image cache; But maybe i can test the performance between the draw cache image and  repaly every single draw when the first CanvasElement has been drawn.